### PR TITLE
doc/rados: edit balancer.rst

### PR DESCRIPTION
**Pull request by zdover23**
Edit doc/rados/operations/balancer.rst.

https://tracker.ceph.com/issues/58485

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/rados/operations/balancer.rst [hunk 1]
```diff
@@ -3,14 +3,15 @@
 Balancer
 ========
 
-The *balancer* can optimize the placement of PGs across OSDs in
-order to achieve a balanced distribution, either automatically or in a
-supervised fashion.
+The *balancer* can optimize the allocation of placement groups (PGs) across
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:6]: I think underscores fit our pattern better

File: doc/rados/operations/balancer.rst [hunk 2]
```diff
@@ -20,70 +21,78 @@ The current status of the balancer can be checked at any time with:
...
    .. prompt:: bash $
 
       ceph balancer off
 
-The balancer mode can be changed to ``crush-compat`` mode, which is
-backward compatible with older clients, and will make small changes to
-the data distribution over time to ensure that OSDs are equally utilized.
+The balancer mode can be changed from ``upmap`` mode to ``crush-compat`` mode.
+The ``crush-compat`` mode is backward compatible with older clients.  In
+``crush-compat`` mode, the balancer automatically makes small changes to the
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:34]: s/the//

File: doc/rados/operations/balancer.rst [hunk 3]
```diff
@@ -20,70 +21,78 @@ The current status of the balancer can be checked at any time with:
...
+If the cluster is degraded (that is, if an OSD has failed and the system hasn't
+healed itself yet), then the balancer will not make any adjustments to the PG
+distribution.
 
-When the cluster is healthy, the balancer will throttle its changes
-such that the percentage of PGs that are misplaced (i.e., that need to
-be moved) is below a threshold of (by default) 5%.  The
-``target_max_misplaced_ratio`` threshold can be adjusted with:
+If the cluster is healthy, then the balancer will throttle its changes.  More
+specifically, the percentage of PGs that are misplaced and that need to be
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:46]: Seems even more stilted this way.  Suggest "When the cluster is healthy, the balancer will incrementally move a small fraction of unbalanced PGs in order to improve distribution.  This fraction will not exceed a certain threshold that deafaults to 5%."

File: doc/rados/operations/balancer.rst [hunk 4]
```diff
@@ -20,70 +21,78 @@ The current status of the balancer can be checked at any time with:
...
+moved will be kept below a certain threshold (default: 5%). To adjust this
+``target_max_misplaced_ratio`` threshold setting, run the following command:
 
    .. prompt:: bash $
 
       ceph config set mgr target_max_misplaced_ratio .07   # 7%
 
-Set the number of seconds to sleep in between runs of the automatic balancer:
+The automatic balancer sleeps between its runs. To set the number of seconds
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:54]: The balancer sleeps between runs.  To set the number length in seconds of this interval, run

File: doc/rados/operations/balancer.rst [hunk 5]
```diff
@@ -20,70 +21,78 @@ The current status of the balancer can be checked at any time with:
...
+for this period of sleep, run the following command:
 
    .. prompt:: bash $
 
       ceph config set mgr mgr/balancer/sleep_interval 60
 
-Set the time of day to begin automatic balancing in HHMM format:
+To set the time of day (in HHMM format) at which automatic balancing begins,
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:61]: s/automatic//

IMHO it's redundant here and below.  ymmv.

File: doc/rados/operations/balancer.rst [hunk 6]
```diff
@@ -93,43 +102,41 @@ The numeric pool IDs can be gotten with the :command:`ceph osd pool ls detail` c
...
-   weights for devices in the CRUSH hierarchy.  The normal weights
-   should remain set to the size of the device to reflect the target
-   amount of data that we want to store on the device.  The balancer
-   then optimizes the weight-set values, adjusting them up or down in
-   small increments, in order to achieve a distribution that matches
-   the target distribution as closely as possible.  (Because PG
-   placement is a pseudorandom process, there is a natural amount of
-   variation in the placement; by optimizing the weights we
-   counter-act that natural variation.)
+#. **crush-compat**. The CRUSH compat mode uses the compat weight-set feature
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:107]: This mode uses

File: doc/rados/operations/balancer.rst [hunk 7]
```diff
@@ -93,43 +102,41 @@ The numeric pool IDs can be gotten with the :command:`ceph osd pool ls detail` c
...
+   (introduced in Luminous) to manage an alternative set of weights for devices
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:108]: I think we s/alternative/shadow/ elswhere

File: doc/rados/operations/balancer.rst [hunk 8]
```diff
@@ -93,43 +102,41 @@ The numeric pool IDs can be gotten with the :command:`ceph osd pool ls detail` c
...
+   in small increments, in order to achieve a distribution that matches the
+   target distribution as closely as possible. (Because PG placement is a
+   pseudorandom process, it is subject to a natural amount of variation;
+   optimizing the weights serves to counteract that natural variation.)
 
-   Notably, this mode is *fully backwards compatible* with older
-   clients: when an OSDMap and CRUSH map is shared with older clients,
-   we present the optimized weights as the "real" weights.
+   Note that this mode is *fully backward compatible* with older clients: when
+   an OSDMap and a CRUSH map are shared with older clients, Ceph presents the
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:119]: the OSD map and CRUSH map

File: doc/rados/operations/balancer.rst [hunk 9]
```diff
@@ -93,43 +102,41 @@ The numeric pool IDs can be gotten with the :command:`ceph osd pool ls detail` c
...
-   handle multiple CRUSH hierarchies with different placement rules if
-   the subtrees of the hierarchy share any OSDs.  (This is normally
-   not the case, and is generally not a recommended configuration
-   because it is hard to manage the space utilization on the shared
-   OSDs.)
+   The primary limitation of this mode is that the balancer cannot handle
+   multiple CRUSH hierarchies with different placement rules if the subtrees of
+   the hierarchy share any OSDs. (Such sharing of OSDs is not normal and,
+   because of the difficulty of managing the space utilization on the shared
+   OSDs, is generally not recommended.)
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:126]: s/normal/typical/

File: doc/rados/operations/balancer.rst [hunk 10]
```diff
@@ -93,43 +102,41 @@ The numeric pool IDs can be gotten with the :command:`ceph osd pool ls detail` c
...
-   placement of individual PGs in order to achieve a balanced
-   distribution.  In most cases, this distribution is "perfect," which
-   an equal number of PGs on each OSD (+/-1 PG, since they might not
-   divide evenly).
+#. **upmap**. In Luminous and later releases, the OSDMap can store explicit
+   mappings for individual OSDs as exceptions to the normal CRUSH placement
+   calculation. These ``upmap`` entries provide fine-grained control over the
+   PG mapping. This balancer mode optimizes the placement of individual PGs in
+   order to achieve a balanced distribution.  In most cases, the resulting
+   distribution is "perfect": that is, there is an equal number of PGs on each
```
Comment (anthonyeleven) [inline doc/rados/operations/balancer.rst:133]: is nearly perfect

**Review by anthonyeleven — APPROVED** [review]
You know the drill

File: doc/rados/operations/balancer.rst (hunk 1 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:6]: RST can't produce underscores. We have bold and italics.

File: doc/rados/operations/balancer.rst (hunk 3 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:46]: Yep. Your suggested text is much more natural.

Comment (zdover23) [reply doc/rados/operations/balancer.rst:46]: Accepted (obviously).

File: doc/rados/operations/balancer.rst (hunk 4 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:54]: Accepted.

File: doc/rados/operations/balancer.rst (hunk 5 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:61]: I get your point, but the docs refer to the "automatic balancer". I'll put this on a list of stuff to rectify later in the year, when we're tightening up the terminology. Fear not: you will be consulted during this great reconciliation.

File: doc/rados/operations/balancer.rst (hunk 6 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:107]: Oh, alright. This change was 100 percent my prejudice for sentences that can be quoted out of context, but I bow to the insight of your fresh eyes.

File: doc/rados/operations/balancer.rst (hunk 7 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:108]: I'll bear this in mind for the great terminological rectification. I'm going to let it stand for now.

File: doc/rados/operations/balancer.rst (hunk 8 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:119]: oof. Good catch.

File: doc/rados/operations/balancer.rst (hunk 9 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:126]: Yeah, that is unambiguously better. Good call.

File: doc/rados/operations/balancer.rst (hunk 10 above)
Comment (zdover23) [reply doc/rados/operations/balancer.rst:133]: Okay. Accepted.

**Comment by zdover23** [comment]
jenkins test submodules

File: doc/rados/operations/balancer.rst (hunk 9 above)
Comment (anthonyeleven) [reply doc/rados/operations/balancer.rst:126]: I started off changing `hierarchies` to `roots`, but realized that there are situations where this comes into play under a single root too, sometimes people make fake duplicate hosts, and the way Rook rolls makes the `.mgr` pool problematic until modified.

File: doc/rados/operations/balancer.rst (hunk 6 above)
Comment (anthonyeleven) [reply doc/rados/operations/balancer.rst]: There's value in that sometimes.  Here " c r u s h - c o m p a t " is a literal and the CRUSH acronym doesn't really fit, that's really what I was going for.

File: doc/rados/operations/balancer.rst (hunk 5 above)
Comment (anthonyeleven) [reply doc/rados/operations/balancer.rst:61]: Fair enough, when I see "automatic balancer" that implies to me that there's at least one other.  Like when Jay Lake said he was going home to his beautiful wife and I asked how many he had.

File: doc/rados/operations/balancer.rst (hunk 2 above)
Comment (anthonyeleven) [reply doc/rados/operations/balancer.rst:34]: The internal struggle between text that flows and grammatical rigor is real and forever, like Cenobites or Lazarus and anti-Lazarus.

File: doc/rados/operations/balancer.rst (hunk 1 above)
Comment (anthonyeleven) [reply doc/rados/operations/balancer.rst:6]: I meant the markup \_balancer\_

**Comment by zdover23** [comment]
jenkins retest this please

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/51824 - Reef backport
https://github.com/ceph/ceph/pull/51825 - Quincy backport
https://github.com/ceph/ceph/pull/51826 - Pacific backport

**Pull request merged**